### ISSUE TITLE
Dispnet demo results view

### ISSUE THREAD

**Issue by varteev** [lmb-freiburg/dispnet-flownet-docker#5]
Hello,
I have downloaded dispflownet-release-1.2 and want to try DispNet demo. I have build caffe as written in readme and run demo.py. After script executed I can see several dispnet-pred-XXX.pfm files (by the way script says "The resulting disparity is stored in dispnet-pred-NNNNNNN.flo"). How can I view pfm files? Is it possible to convert them to JPG or some format to visualise results?

**Comment by nikolausmayer** [comment]
We have Python I/O code for PFM online, search [here](https://lmb.informatik.uni-freiburg.de/resources/datasets/SceneFlowDatasets.en.html) for "snippet".
For C++, I recommend [CImg](http://cimg.eu).

You can use these to read PFM files, then display or save as JPG. Note that the pixel range is not the (0,255) that color images expect, so you'll need to rescale the data before saving.

Thanks for the FLO/PFM hint ;)

**Comment by varteev** [comment]
Thanks a lot! Python snippet works.

**Comment by OzzyW** [comment]
Is it possible that the pixel range is different for DispNet and for DispNetCorr1D-K which is actually in the (0,255) range?

**Comment by nikolausmayer** [comment]
@OzzyW AFAIK all model prototxt files should contain "Eltwise" layers (at the beginning) with "coeff: 0.00392...". This is just "division by 255".
EDIT: Wait, you're talking about the outputs I guess. I don't think the networks contain anything to guarantee that range, but it depends on your inputs of course. I did not want to say that "the range is different"; what I meant is that if a disparity value >255 appears, that range will not suffice to capture it...

**Comment by OzzyW** [comment]
Yeah that is what I meant. 
I guess that best solution is to use CImg and convert to another more "friendly" format.

**Comment by nikolausmayer** [comment]
I'd like to add something, just in case it may be causing confusion: Unlike PGM/PPM or even 8-bit or 16-bit PNG, PFM has no tighter value range than [FLOAT_MIN, FLOAT_MAX]. 
PFM is not really an "image" format, it's more a general data storage that happens to be single- or triple-channel. One should not expect to be able to "display" a PFM without doing value normalization and range adjustment.

**Comment by OzzyW** [comment]
I believe the Python snippet works well only for your provided DispNetCorr1D-K caffemodel (as the range seems to be normalized and adjusted). How exactly do I normalize and range adjust as my trained output results differ? If I try the snippet on DispNet it is not normalized or adjusted the same. 
In the deploy prototext there are 2 layers which I think are responsible for this (DataAugmentation and Resample). Is configuring them and the demo script the way to go or am I off?

Edit: OK I think it just boils down to the input used. But anyways what are the DataAugmentation and Resample layers used  for?

**Comment by nikolausmayer** [comment]
The "deploy" prototext file templates for all shipped DispNet variants (DispNet, DispNet-K, DispNetCorr1D, DispNetCorr1D-K) contain:
- "Eltwise" layers in the beginning which transform (0,255) inputs into (0,1) data
- "Augmentation" layers for demeaning
- "Resample" layers in the beginning and end. These are necessary because the network needs multiples-of-64 data dimensions. The resampling layers in the beginning upsample the inputs, and the resampling layers in the end downsample the output to the input size.
- "Eltwise" layers in the end which scale the output to correct the mismatch between the original input size and the upsampled data which is actually processed.

There really should not be a difference in the ranges produced by the networks, and we do not perform any normalization or scaling of the final network outputs. After all, (0,255) has no connection to the data range of the disparities.

But if a sample happens to produce a value outside of the range (0,255) in one network and not in the other, converting that output into a (0,255) format without explicitly clipping or rescaling will result in bad pixels. 

Could you post examples of how your outputs for DispNetCorr1D-K and DispNet differ?

**Comment by OzzyW** [comment]
Here is an example of the same image tested:

DispNetCorr1D-K
![dk](https://user-images.githubusercontent.com/21277219/41045146-0bc77f70-69a8-11e8-9738-b3215f0119c6.png)

DispNet
![dispnet](https://user-images.githubusercontent.com/21277219/41045137-07a44126-69a8-11e8-821b-f19bea502d85.png)

Just KITTI
![k](https://user-images.githubusercontent.com/21277219/41045161-14049f10-69a8-11e8-8dfb-2a32ae164062.png)

So I am in the range of (0,255) only in the DispNetCorr1D-K and that is problematic to compare them to each other using the same Python snippet.

And just to understand better:
1) Why is moving to the  (0,1) range better?
2) Demeaning improves the accuracy (as I have seen other networks which do not do this)?
3) "The upsampled data which is actually processed." - you mean processed by the network itself?

And thanks once more, I really appreciate your help :1st_place_medal:

**Comment by nikolausmayer** [comment]
The first two results are as I would expect them: They contain roughly the same values, but the KITTI-finetuned network can deal with the street better and predicts (correctly) larger values. The second network has a larger maximum, but that's likely some spurious wrong match _somewhere_ in the image. 
We can also observe that it predicts negative values which are not possible in reality: the network is "confused" by something in this input. But that's ok, or at least not unexpected.

Your "just KITTI" network's output is strange. It predicts huge values at the bottom, and the values of horizontally neighboring pixels vary by >1px.

Considering that the DispNetCorr1D-K is pretty good on KITTI-style scenes, my first guess would be that your "just KITTI" net receives wrongly scaled inputs, or that there is a difference between its training and testing that does not exist in our nets...

**Comment by OzzyW** [comment]
I agree. How did you input the KITTI in for the fine tuning on DispNetCorr1D-K? I just took the raw data they provided and plugged it in for the "just KITTI" and it must be different from the way you did it.

But why are the **trained** DispNets  scaled differently from yours as I did not input anything differently? (I copied the prototext files from the DispNet for DispNetCorr1D-K training and did not yet train on any KITTI data on the DispNetCorr1D-K)

My trained DispNetCorr1D-K 600K
![600kdk](https://user-images.githubusercontent.com/21277219/41097722-4739bd46-6a59-11e8-81e0-b765512c6d54.png)

My trained DispNetCorr1D-K 1200K
![1200kdk](https://user-images.githubusercontent.com/21277219/41097725-4b026356-6a59-11e8-8b65-faf833be43a6.png)

My trained DispNet 600K (did not continue to train)
![600kd](https://user-images.githubusercontent.com/21277219/41097731-4dda5f84-6a59-11e8-9a5c-36890a07e440.png)

**Comment by nikolausmayer** [comment]
We only used the KITTI 2012/2015 training datasets for finetuning, not KITTI-raw. We converted the PNG groundtruth to PFM, replaced zeroes with NaNs and then used it the same way as e.g. FlyingThings3D.

Can you elaborate what you mean by "I copied the prototext files from the DispNet for DispNetCorr1D-K training"? DispNet and DispNetCorr1D are different architectures, and their prototext files are mutually incompatible.

**Comment by OzzyW** [comment]
I meant raw in that sense that nothing was changed, we used the same datasets in KITTI. :)

In your first comment here https://github.com/lmb-freiburg/dispnet-flownet-docker/issues/4 you said: "The train.prototxt are just the same as for the "non-K" nets." So I used the same prototexts (as there are none available for the DispNetCorr1D-K) to train and deploy both nets. Any idea what could cause this difference in the results?

**Comment by nikolausmayer** [comment]
That #4 comment meant that "DispNet" and "DispNet-K" have the same prototext files, and "DispNetCorr1D" and "DispNetCorr1D-K" as well. But they are not interchangeable across pairs, i.e. a correlation-net prototext file will not work for a net without correlation layer and vice versa.

Is it possible that your networks are terribly overfitting? Are the samples you posted from the training set or from a different sample?

Have you looked at how your loss behaves? The latest results you posted look so totally wrong (no offense :wink:) that I can't help wonder if your networks are converging at all.


EDIT: No, the image you posted in #7 shows that your network is converging and not overfitting too badly... there goes my last idea.

**Comment by OzzyW** [comment]
Oh, that makes sense now but even if we ignore the DispNetCorr1D-K training I still do not understand why does the scaling comes out different? :( My posted samples are from an untrained different sample. The just KITTI did have really big losses and didn't seem to converge but the DispNet seemed to be OK IIRC (but both seem to be normalized as seen in https://github.com/lmb-freiburg/dispnet-flownet-docker/issues/7).

**Comment by nikolausmayer** [comment]
There really is no normalization in #7, and none of our networks normalize their outputs. It is coincidence that KITTI’s disparities are not larger than 255. If a network does badly on a sample, it can produce larger values.
But I agree that your networks seem to have a scaling problem. I just can't think of a reason why... :disappointed:

**Comment by OzzyW** [comment]
Was the relevant layer's loss_weight turned on/off progressively based on iteration count?

**Comment by nikolausmayer** [comment]
Yes; there are 7 breakpoints at which the loss weights switch:
`[0 iterations, 50k, 100k, 150k, 250k, 350k, 450k]`

This table gives the loss weights per loss layer. For example, loss3 switches to weight 0.2 after the third breakpoint (100k iterations). The loss names should match those in the train.prototxt files in our code release.
```
loss6: [1.0, 0.5, 0, 0, 0, 0, 0]
loss5: [0.2, 1, 0.5, 0, 0, 0, 0]
loss4: [0, 0.2, 1, 0.5, 0, 0, 0]
loss3: [0, 0, 0.2, 1, 0.5, 0, 0]
loss2: [0, 0, 0, 0.2, 1, 0.5, 0]
loss1: [0, 0, 0, 0, 0.2, 1.0, 1.0] (highest trained resolution)
```

Note that during finetuning, IIRC only the last loss is active.

**Comment by OzzyW** [comment]
Perfect! I will check if this may be the reason the results are in a different scale.

**Comment by OzzyW** [comment]
The scale is changing only a bit. When I tried to look at the loss of your trained caffemodel the losses are totally different then when I train them. Once the net gets my input it causes the loss to increase and then decrease immediately. My guess it is something to do with the input but I did use the make-lmdbs.sh which calls the convert_imageset_and_disparity so not sure where something goes wrong...

When I train from scratch:
![mine](https://user-images.githubusercontent.com/21277219/41533461-200a067c-72fb-11e8-9b2c-acde0d5ab462.png)

When I continue training provided caffemodel:
![his](https://user-images.githubusercontent.com/21277219/41533467-23eb6fce-72fb-11e8-8f39-7ed3fc044de2.png)

**Comment by nikolausmayer** [comment]
Are only the last (lowest-index/highest-resolution) losses enabled in these examples? E.g. the last sample in our pretrained caffemodel shows a high loss at level 5 but a low loss at level 6. It looks a bit weird, but I guess it could be ok since these layers have been trained without attached loss for a long time.

**Comment by OzzyW** [comment]
Yes but in mine I let it train before in a similar way to the loss weights switch you mentioned. It is a bit weird but goes down as training continues (I do not know why for example level 6 changes immediately though). I was thinking of changing the last "Eltwise" layer but this doesn't make much sense, does it?

**Comment by nikolausmayer** [comment]
Uh, now I'm confused. Which Eltwise layer are we talking about? The only `Eltwise` layers in our training files should be the ones that scale the initial input to [0,1] data range.
Do you mean the result-rescaling layer in the `deploy` file?

**Comment by OzzyW** [comment]
Sorry for the confusion. I was talking about the deploy file in order to scale the output but there **must** be a better solution we have not found yet :/

**Comment by nikolausmayer** [comment]
Ah, I see. Yes I agree, but I'm also confused because there really isn't that much complexity in the model. It's not like we have a wildly different training process on our side than in the released code. And the pretrained models work, so that's not it either. I assume you are not observing any error messages from Caffe when you load the models for continued training?

**Comment by OzzyW** [comment]
I only had some warnings in the compile but nothing major. I have no errors running it and it works great for the pretrained models so have no idea what causes this yet

**Comment by adhara123007** [comment]
For the loss schedule you listed above,
loss6: [1.0, 0.5, 0, 0, 0, 0, 0]
loss5: [0.2, 1, 0.5, 0, 0, 0, 0]
loss4: [0, 0.2, 1, 0.5, 0, 0, 0]
loss3: [0, 0, 0.2, 1, 0.5, 0, 0]
loss2: [0, 0, 0, 0.2, 1, 0.5, 0]
loss1: [0, 0, 0, 0, 0.2, 1.0, 1.0] (highest trained resolution)

Did you use the same solver.prototxt file that is present in the folder?

net: "../model/train.prototxt"
test_iter: 4370
test_interval: 10000
base_lr: 1e-4
lr_policy: "multistep"
gamma: 0.5
stepvalue: 200000
stepvalue: 350000
stepvalue: 500000
stepvalue: 650000
stepvalue: 800000
momentum: 0.9
weight_decay: 0.0004
display: 50
# Note: we initialized this net with pre-trained weights from another net and stopped at iteration 620000 
# Also, to faster train the network we used loss weight schedules (see paper), which are not included here 
max_iter:  1200000
snapshot: 20000
snapshot_prefix: "disp"
solver_mode: GPU
solver_type: ADAM
momentum2: 0.999
delta: 1e-4
